### PR TITLE
Ensure that TopicReplicaDistributionGoal balances immigrant replicas during self healing and uses a balance margin

### PR DESCRIPTION
**Pull request by efeg**
Addresses the issues https://github.com/linkedin/cruise-control/issues/600 and https://github.com/linkedin/cruise-control/issues/624.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java [hunk 1]
```diff
@@ -290,22 +333,147 @@ private void healCluster(ClusterModel clusterModel,
...
+
+  private boolean rebalanceByMovingReplicasIn(Broker broker,
+                                              String topic,
+                                              ClusterModel clusterModel,
+                                              Set<Goal> optimizedGoals,
+                                              OptimizationOptions optimizationOptions) {
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+
+    PriorityQueue<Broker> eligibleBrokers = new PriorityQueue<>((b1, b2) -> {
+      int result = Double.compare(b2.numReplicasOfTopicInBroker(topic), b1.numReplicasOfTopicInBroker(topic));
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java]: `Double.compare` -> `Integer.compare`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java [hunk 2]
```diff
@@ -290,22 +333,147 @@ private void healCluster(ClusterModel clusterModel,
...
+    }
+
+    Set<Broker> candidateBrokers = Collections.singleton(broker);
+    int numReplicasOfTopicInBroker = broker.numReplicasOfTopicInBroker(topic);
+
+    // Stop when no topic replicas can be moved in anymore.
+    while (!eligibleBrokers.isEmpty()) {
+      Broker sourceBroker = eligibleBrokers.poll();
+      SortedSet<Replica> replicasToMove = replicasToMoveOut(clusterModel, sourceBroker, topic);
+
```
Comment (kidkun) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java]: redundant empty line.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/BalancingConstraint.java [hunk 3]
```diff
@@ -96,12 +100,19 @@ public Long maxReplicasPerBroker() {
...
   /**
-   * Get replica balance percentage for replica distribution goal.
+   * Get replica balance percentage for {@link com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaDistributionGoal}.
    */
   public Double replicaBalancePercentage() {
     return _replicaBalancePercentage;
   }
 
+  /**
+   * Get topic replica balance percentage for {@link TopicReplicaDistributionGoal}.
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/BalancingConstraint.java]: didn't notice that -- thanks for pointing out!

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java [hunk 4]
```diff
@@ -160,9 +162,20 @@ public Replica replica(TopicPartition tp) {
    * @return Replicas in this broker sharing the given topic.
    */
   public Collection<Replica> replicasOfTopicInBroker(String topic) {
-    Map<Integer, Replica> topicReplicas = _topicReplicas.get(topic);
+    SortedMap<Integer, Replica> topicReplicas = _topicReplicas.get(topic);
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java]: We try to keep the number of spaces between two function implementations as 1 line -- other than  this, I feel that a single space between such a statement makes the code more clean.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java [hunk 5]
```diff
@@ -49,7 +51,7 @@
   /** Set of immigrant replicas */
   private final Set<Replica> _immigrantReplicas;
   /** A map for tracking topic -&gt; (partitionId -&gt; replica). */
-  private final Map<String, Map<Integer, Replica>> _topicReplicas;
+  private final Map<String, SortedMap<Integer, Replica>> _topicReplicas;
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Broker.java]: Correct, this change is no longer needed -- thanks!

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java [hunk 6]
```diff
@@ -290,22 +333,147 @@ private void healCluster(ClusterModel clusterModel,
...
+    candidateBrokers.addAll(_selfHealingDeadBrokersOnly ? clusterModel.aliveBrokers() : clusterModel
+        .aliveBrokers()
+        .stream()
+        .filter(b -> b.numReplicasOfTopicInBroker(topic) < _balanceUpperLimitByTopic.get(topic))
+        .collect(Collectors.toSet()));
+
+    int numReplicasOfTopicInBroker = broker.numReplicasOfTopicInBroker(topic);
+
+    for (Replica replica : replicasToMoveOut(clusterModel, broker, topic)) {
+      if (shouldExclude(replica, excludedTopics)) {
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java]: `_avgTopicReplicasOnAliveBroker` is populated per `topic`, but not per `broker`. Hence, it is possible that if there are self healing eligible replicas, their topic could be in the list due to an offline replica from that topic, but another replica from the same topic might be online. This check ensures that we can exclude such replicas.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java [hunk 7]
```diff
@@ -290,22 +333,147 @@ private void healCluster(ClusterModel clusterModel,
...
+                                               String topic,
+                                               ClusterModel clusterModel,
+                                               Set<Goal> optimizedGoals,
+                                               OptimizationOptions optimizationOptions) {
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+    // Get the eligible brokers.
+    SortedSet<Broker> candidateBrokers = new TreeSet<>(
+        Comparator.comparingInt((Broker b) -> b.numReplicasOfTopicInBroker(topic)).thenComparingInt(Broker::id));
+
+    candidateBrokers.addAll(_selfHealingDeadBrokersOnly ? clusterModel.aliveBrokers() : clusterModel
```
[the remark this answers is not in the document]
Comment (efeg) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java]: The current implementation provides a single, clean path for all goals. This seems to be a general suggestion that applies to all goals.

Would you please create a ticket for this optimization proposal?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (kidkun) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/TopicReplicaDistributionGoal.java]: ah, you are right, it is per cluster.

**Pull request merged**